agasc/supplement/utils.py

```diff
@@ -134,3 +144,191 @@ def save_version(filename, table_name):
+    :param dry_run: bool
+    :param create: bool
+        Create a supplement file if it does not exist
+    """
+    filename = Path(filename)
+
+    if len(table) == 0:
+        logger.info('Nothing to update')
+        return
+
+    suppl_table = get_table(filename, path, dtype, create=create)
+    table = Table(table if len(table) else None, dtype=dtype)
+    table = unique(table, keys=keys, keep='first')
+
+    # for now I require no masked elements.
+    if table.mask is not None and np.any(table.mask):
+        masked = [name for name in table.colnames if np.any(table[name].mask)]
+        raise Exception('"{path}" table should have no masked elements, '
+                        'but element in these columns are masked: ',
+                        ', '.join(masked))
+
+    # entries already in supplement
+    intersect = (table[keys[0]][None] == suppl_table[keys[0]][:, None])
+    for key in keys[1:]:
+        intersect &= (table[key][None] == suppl_table[key][:, None])
+    i, j = np.argwhere(intersect).T
+
+    # entries not yet in supplement
+    append = (~np.in1d(table[keys[0]], suppl_table[keys[0]]))
+    for key in keys[1:]:
+        append |= (~np.in1d(table[key], suppl_table[key]))
+    k = np.argwhere(append).T
+
+    if not len(i) and not len(k[0]):
+        return
+
+    logger.info(f'updating "{path}" table in {filename}')
+
+    if len(i):
+        suppl_table[i] = table[j]
+    if len(k[0]):
+        suppl_table = vstack([suppl_table, table[k[0]]])
+
+    if not dry_run:
+        suppl_table.write(str(filename), format='hdf5', path=path, append=True, overwrite=True)
+        save_version(filename, path)
+    else:
+        logger.info('dry run, not saving anything')
+
+
+def get_table(filename, path, dtype=None, create=False):
+    """
+    Gets a table from an HDF5 file. Optionally creates the file if it does not exist.
+    Creates the table if it does not exist.
+
+
+    :param filename:
+    :param path: str
+        the path of the table in the supplement ('obs', 'mags')
+    :param dtype: np.dtype
+        the dtype of the table. Used only if the table is not in the supplement already.
+        If it is not given, and the table does not exist in the supplement,
+        an exception will be raised.
+    :param create: bool
+        Create a supplement file if it does not exist
+    """
+    if not Path(filename).exists():
+        if not create:
+            raise FileExistsError(filename)
+        logger.warning(f'Creating a new AGASC supplement: {filename}')
+
+    try:
+        result = Table.read(filename, format='hdf5', path=path)
+    except OSError as e:
+        if not filename.exists() or str(e) == f'Path {path} does not exist':
+            result = Table(dtype=dtype)
+        else:
+            raise
+
+    return result
+
+
+def add_bad_star(filename, bad, dry_run=False, create=False):
+    """
+    Update the 'bad' table of the AGASC supplement.
+
+    :param filename: pathlib.Path
```

Comment:
Presumably this can also be `str`.  This and the rest of these args need descriptions. (This comment applies to several other functions as well).

Reply:
I improved the docstrings a bit.